File: test/conformance/java/rows/F10.java
// test/conformance/java/rows/F10.java
//
// cell java/F10 — "no_layout" (docs/FIXED-FORM-ALGORITHM.md §5.3 step 7; the
// schema matrix's plan-selection row). A fixed-form file whose per-record hash
// does not equal the file's own hash is refused BY NAME — no_layout — before
// any body byte is decoded and before any counter moves.
//
// The law (docs/FIXED-FORM-ALGORITHM.md:148):
//   | 7 | per record: if LE(8, at) != h, REFUSE no_layout; then §4.4 |
//
// The vector is built from the law, by the generated writer's own constants
// (build/tables-generated-java/v1, package tblv1, on the conformance
// classpath): the header is the form byte 3, seven reserved zero bytes, the
// layout hash at 8, the layout length at 16 and the layout itself; then one
// record of recordBytes = 8 + bodyBytes, whose first eight bytes are the
// per-record hash. Only that per-record hash is corrupted — a record hash a
// reader holds no layout for is exactly what step 7 refuses.
//
// Run from the repository root against the conformance classpath, the same one
// make/java.mk gives the conformance driver (build/conformance-java):
//   java -cp build/conformance-java test/conformance/java/rows/F10.java
// exit 0 green, exit 1 red, one printed line per assertion.

import tblv1.CellFixed;
import tblv1.TableFixed;

public final class F10 {
    private static int failures = 0;

    private static void check(boolean ok, String line) {
        System.out.println((ok ? "pass" : "FAIL") + ": " + line);
        if (!ok) {
            failures++;
        }
    }

    public static void main(String[] args) {
        // A valid one-record file, built with the generated writer's own
        // constants: header (form 3, reserved zeros, the layout hash at 8, the
        // layout length at 16, the layout) then one record. Its first eight
        // bytes, at `recordsAt`, are the per-record hash (LE(8, at)).
        byte[] file = new byte[CellFixed.measure(1)];
        int recordsAt = CellFixed.writeHeader(file);
        CellFixed.Value v = new CellFixed.Value();
        v.power = 7;
        CellFixed.writeBody(file, recordsAt + 8, v);
        TableFixed.put64(file, recordsAt, CellFixed.hash);

        // Step 7: if LE(8, at) != h, REFUSE no_layout. Flip one byte of the
        // record's hash so it names a layout this reader does not hold.
        file[recordsAt] ^= 0x01;

        CellFixed.Value[] values = new CellFixed.Value[1];
        values[0] = new CellFixed.Value();
        TableFixed.Report r = new TableFixed.Report();
        int n = CellFixed.load(values, 1, file,
                CellFixed.identityPlan(), new short[8], CellFixed.image(), r);

        check(n == -1, "a record whose hash names no layout is refused (load returned " + n + ")");
        check(r.refused, "the load refuses");
        check(r.reason == TableFixed.Reason.noLayout,
                "the refusal is by the name noLayout (got " + r.reason + ")");
        check(!r.malformed, "noLayout is a refusal, not malformed");
        check(r.unknown == 0 && r.kindMismatch == 0 && r.widened == 0 && r.clamped == 0,
                "REFUSE is total: no counter moves");
        check(values[0].power == 0, "nothing decoded: the destination is untouched");

        System.exit(failures == 0 ? 0 : 1);
    }
}

File: test/conformance/c/rows/F10.c
/* F10 — no_layout, the C leg of the schema matrix cell (docs/roadmap.sexp
   row group "plan-selection"; docs/FIXED-FORM-ALGORITHM.md:148 step 7):

       "per record: if `LE(8, at) != h`, `REFUSE no_layout`; then §4.4"

   A FIXED-form record whose per-record hash is not the file's own header hash
   refuses under the NAME `no_layout`, before one destination byte is written.
   The header hash is untouched, so the file still selects this build's own
   known layout — only the record's hash is forged, which is the no_layout
   law and not layout_newer (hash outside the lineage) or layout_malformed
   (a known hash over different layout bytes).

   VECTOR DERIVATION. A clean one-record V1 `Cell` file is what the generated
   writer itself lays down; `cell_fixed_save` stamps every record's per-record
   hash with the header's own hash, LE-encoded. That record hash sits at:

       header (kTableFixedHeaderBytes = 16)
     + the layout's u32 length (4)
     + the layout itself (cell_fixed_layout_bytes)
     = rec_at

   Inverting one byte there makes LE(8, at) != h while every earlier check
   (form byte, reserved bytes, layout length, hash selection, layout memcmp,
   tail a whole record) still passes, so the load reaches the per-record hash
   check and refuses there. The caller's storage is poisoned 0x5A first:
   REFUSE is total (§5.8), so no counter may move and no byte may be written.

   SELF-CONTAINED: reaches the generated fixed form through the SAME include
   path the conformance driver uses (build/tables-generated-c/v1), includes
   only V1Table.h, and depends on no other rows/ file and no shared fixture. */

#include <stdio.h>
#include <stdint.h>
#include <string.h>

#include "V1Table.h"

int main( void )
{
    Cell source;
    Cell back;
    TableReport r;
    uint8_t wire[256];
    int64_t bytes, rec_at, i, n;

    cell_reset( &source );
    source.power = 7;
    memcpy( source.label, "hello", 6 );
    source.label_length = 5;

    bytes = cell_fixed_measure( 1 );
    if ( bytes < 0 || bytes > (int64_t) sizeof( wire ) )
    {
        printf( "F10 no_layout: measure out of range: %lld\n", (long long) bytes );
        return 1;
    }
    if ( cell_fixed_save( &source, 1, wire, bytes ) != bytes )
    {
        printf( "F10 no_layout: the clean save failed\n" );
        return 1;
    }

    /* the forge: record 0's own hash, one byte inverted */
    rec_at = kTableFixedHeaderBytes + 4 + cell_fixed_layout_bytes;
    wire[rec_at] ^= 0xff;

    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );

    n = cell_fixed_load( &back, 1, wire, bytes, NULL, 0, NULL, &r );

    if ( n != -1 )
    {
        printf( "F10 no_layout: the forged record hash did not refuse: n=%lld\n", (long long) n );
        return 1;
    }
    if ( !r.refused )
    {
        printf( "F10 no_layout: refused not set\n" );
        return 1;
    }
    if ( r.reason != SCHEMA_TABLE_NO_LAYOUT )
    {
        printf( "F10 no_layout: reason=%d, not no_layout\n", r.reason );
        return 1;
    }
    if ( r.malformed )
    {
        printf( "F10 no_layout: malformed set — a refusal by name is not a malformed file\n" );
        return 1;
    }
    if ( r.clamped || r.unknown || r.kind_mismatch || r.widened || r.duplicate )
    {
        printf( "F10 no_layout: a counter moved: clamped=%d unknown=%d kind_mismatch=%d widened=%d duplicate=%d\n",
                r.clamped, r.unknown, r.kind_mismatch, r.widened, r.duplicate );
        return 1;
    }
    for ( i = 0; i < (int64_t) sizeof( back ); ++i )
    {
        if ( ( (const uint8_t *) &back )[i] != 0x5A )
        {
            printf( "F10 no_layout: byte %lld is 0x%02x — REFUSE wrote a destination byte\n",
                    (long long) i, ( (const uint8_t *) &back )[i] );
            return 1;
        }
    }

    printf( "F10 no_layout: a record whose hash names no held layout refuses no_layout; nothing decoded, no byte written\n" );
    return 0;
}
